### ISSUE TITLE
error while debugging threads->create()

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10866]
Migrated from [rt.perl.org#79996](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=79996) (status was 'resolved')

Searchable as RT79996$

**Comment by p5pRT** [comment]
### From thecrux@gmail.com

#### Created by thecrux@gmail.com



While debuging threaded app\, debugger returned unexpected error on
threads\->create\(\) if argument is 'sub \{\.\.\}' or '\\&somesub'

simple testcase&#8203;:

$ PERLDB\_OPTS="noTTY" perl \-Mthreads \-d \-e "threads\->create\(sub\{\}\)\->join"
Thread 1 terminated abnormally&#8203;: Undefined subroutine &Term&#8203;::ReadLine&#8203;::CLONE called at \-e line 1\.
 at \-e line 1 thread 1

no errors in this testcase&#8203;:

$ PERLDB\_OPTS="noTTY" perl \-Mthreads \-d \-e "$r=sub\{\}; threads\->create\($r\)\->join"
$

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=medium
    module=threads

Site configuration information for perl 5.12.2:

Configured by builder at Wed Nov 17 16:02:15 UTC 2010.

Summary of my perl5 (revision 5 version 12 subversion 2) configuration:
   
  Platform:
    osname=linux, osvers=2.6.18-ovz-rhel-alt13.m51.15, archname=x86_64-linux-thread-multi
    uname='linux apiary.egro.altlinux.org 2.6.18-ovz-rhel-alt13.m51.15 #1 smp tue oct 12 07:57:28 utc 2010 x86_64 gnulinux '
    config_args='-ders -Dusethreads -Duseithreads -Duselargefiles -Duseshrplib -Dlibperl=libperl-5.12.so -Dcc=gcc -Doptimize=-pipe -Wall -O2 -Dprefix=/usr -Dprivlib=/usr/share/perl5 -Darchlib=/usr/lib64/perl5 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib64/perl5 -Dotherlibdirs=/etc/perl5:/usr/lib/perl5/vendor_perl -Dinc_version_list=none -Dpager=/usr/bin/less -isR -Dman1dir=none -Dman3dir=none -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.12.2 -Dsitearch=/usr/local/lib64/perl/5.12.2'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-pipe -Wall -O2',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.5.1 20101112 (ALT Linux 4.5.1-alt6)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64 /usr/local/lib64
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lpthread -lc -lgdbm_compat
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=, so=so, useshrplib=true, libperl=libperl-5.12.so
    gnulibc_version='2.11.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -pipe -Wall -O2 -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.12.2:
    /usr/local/lib64/perl/5.12.2
    /usr/local/share/perl/5.12.2
    /usr/lib64/perl5
    /usr/share/perl5
    /usr/lib64/perl5
    /usr/share/perl5
    /etc/perl5
    /usr/lib/perl5/vendor_perl
    .


Environment for perl 5.12.2:
    HOME=/home/crux
    LANG=ru_RU.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/crux/bin:/bin:/usr/bin:/usr/local/bin:/usr/X11R6/bin:/usr/games
    PERL_BADLANG (unset)
    SHELL=/usr/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Nov 29 23&#8203;:25&#8203;:48 2010\, crux wrote&#8203;:
> While debuging threaded app\, debugger returned unexpected error on
> threads\->create\(\) if argument is 'sub \{\.\.\}' or '\\&somesub'
> 
> simple testcase&#8203;:
> 
> $ PERLDB\_OPTS="noTTY" perl \-Mthreads \-d \-e "threads\->create\(sub\{\}\)\-
>    >join"
> Thread 1 terminated abnormally&#8203;: Undefined subroutine
>    &Term&#8203;::ReadLine&#8203;::CLONE called at \-e line 1\.
>  at \-e line 1 thread 1

Thank you for your report\. This appears to have been fixed in bleadperl\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From thecrux@gmail.com

On Sun\, Dec 05\, 2010 at 05&#8203;:25&#8203;:14PM \-0800\, Father Chrysostomos via RT wrote&#8203;:
> On Mon Nov 29 23&#8203;:25&#8203;:48 2010\, crux wrote&#8203;:
> > While debuging threaded app\, debugger returned unexpected error on
> > threads\->create\(\) if argument is 'sub \{\.\.\}' or '\\&somesub'
> > 
> > simple testcase&#8203;:
> > 
> > $ PERLDB\_OPTS="noTTY" perl \-Mthreads \-d \-e "threads\->create\(sub\{\}\)\-
> >    >join"
> > Thread 1 terminated abnormally&#8203;: Undefined subroutine
> >    &Term&#8203;::ReadLine&#8203;::CLONE called at \-e line 1\.
> >  at \-e line 1 thread 1
> 
> Thank you for your report\. This appears to have been fixed in bleadperl\.

Can you point to the commit where this was fixed?
Is it possible to backport this change to 5\.12?

\-\- 
Vladimir Lettiev aka crux ✉ theCrux@&#8203;gmail\.com

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Dec 05 21&#8203;:29&#8203;:22 2010\, crux wrote&#8203;:
> On Sun\, Dec 05\, 2010 at 05&#8203;:25&#8203;:14PM \-0800\, Father Chrysostomos via RT
wrote&#8203;:
> > On Mon Nov 29 23&#8203;:25&#8203;:48 2010\, crux wrote&#8203;:
> > > While debuging threaded app\, debugger returned unexpected error on
> > > threads\->create\(\) if argument is 'sub \{\.\.\}' or '\\&somesub'
> > > 
> > > simple testcase&#8203;:
> > > 
> > > $ PERLDB\_OPTS="noTTY" perl \-Mthreads \-d \-e "threads\->create\(sub\{\}\)\-
> > >    >join"
> > > Thread 1 terminated abnormally&#8203;: Undefined subroutine
> > >    &Term&#8203;::ReadLine&#8203;::CLONE called at \-e line 1\.
> > >  at \-e line 1 thread 1
> > 
> > Thank you for your report\. This appears to have been fixed in bleadperl\.
> 
> Can you point to the commit where this was fixed?
> Is it possible to backport this change to 5\.12?
> 

It was fixed by d08d57ef171\. I think that commit is too intrusive \(and
hence to risky\) to backport to 5\.12\.

**Comment by p5pRT** [comment]
### From thecrux@gmail.com

On Sun\, Dec 12\, 2010 at 01&#8203;:16&#8203;:11PM \-0800\, Father Chrysostomos via RT wrote&#8203;:
> On Sun Dec 05 21&#8203;:29&#8203;:22 2010\, crux wrote&#8203;:
> > On Sun\, Dec 05\, 2010 at 05&#8203;:25&#8203;:14PM \-0800\, Father Chrysostomos via RT
> wrote&#8203;:
> > > On Mon Nov 29 23&#8203;:25&#8203;:48 2010\, crux wrote&#8203;:
> > > > While debuging threaded app\, debugger returned unexpected error on
> > > > threads\->create\(\) if argument is 'sub \{\.\.\}' or '\\&somesub'
> > > > 
> > > > simple testcase&#8203;:
> > > > 
> > > > $ PERLDB\_OPTS="noTTY" perl \-Mthreads \-d \-e "threads\->create\(sub\{\}\)\-
> > > >    >join"
> > > > Thread 1 terminated abnormally&#8203;: Undefined subroutine
> > > >    &Term&#8203;::ReadLine&#8203;::CLONE called at \-e line 1\.
> > > >  at \-e line 1 thread 1
> > > 
> > > Thank you for your report\. This appears to have been fixed in bleadperl\.
> > 
> > Can you point to the commit where this was fixed?
> > Is it possible to backport this change to 5\.12?
> > 
> 
> It was fixed by d08d57ef171\. I think that commit is too intrusive \(and
> hence to risky\) to backport to 5\.12\.

Thank you

\-\- 
Vladimir Lettiev aka crux ✉ theCrux@&#8203;gmail\.com
